### ISSUE TITLE
ACL allows relation access for not found entries even though DENY

### ISSUE THREAD

**Issue by mediastuttgart** [strongloop/loopback#2220]
It SHOULD return `401 AUTHORIZATION_REQUIRED`.

**ex /media/{id}/author** will produce:

````
{
  "error": {
    "name": "Error",
    "status": 404,
    "message": "could not find a model with id x",
    "statusCode": 404,
    "code": "MODEL_NOT_FOUND"
  }
}
````

where author is a relation defined in `media.json`

````
"relations": {
  "author": {
    "type": "belongsTo",
    "model": "profile",
    "foreignKey": "author_id"
  }
}
````
even though the ACL for media is defined as follows

````
"acls": [
  {
    "accessType": "*",
    "principalType": "ROLE",
    "principalId": "$everyone",
    "permission": "DENY"
  }
]
````
even if the same ACL is applied to the author model as well

````
"acls": [
  {
    "accessType": "*",
    "principalType": "ROLE",
    "principalId": "$everyone",
    "permission": "DENY"
  }
]
````

System
````
mongodb: 3.2.3
loopback: 2.26.2
loopback-boot: 2.16.0
loopback-component-explorer: 2.3.0
loopback-connector-mongodb: 1.13.3
loopback-datasource-juggler: 2.45.2
````

**Comment by richardpringle** [comment]
Hey @mediastuttgart, just FYI, if you fork the [loopback-sandbox](https://github.com/strongloop/loopback-sandbox) so I can just clone and reproduce right away it will speed up the process.

Just for future reference, I'll look into this right now though.

**Comment by richardpringle** [comment]
@mediastuttgart I couldn't reproduce, could you set up a sandbox for me please? Might be something pretty specific.

**Comment by mediastuttgart** [comment]
@richardpringle of course. I'll try to set up the sandbox at the weekend.

**Comment by michaelfreund** [comment]
@richardpringle I'm sorry but the issue was not reported correctly.

**The problem is** IF a **not existing model ID** is given, it SHOULD return `401 AUTHORIZATION_REQUIRED`.

Instead **GET /media/{not_existing_id}/author** will return `404 MODEL_NOT_FOUND` with the the following ACL:

`````
    {
      "accessType": "READ",
      "principalType": "ROLE",
      "principalId": "$everyone",
      "permission": "DENY",
      "property": "__get__author"
    }
````

**Comment by michaelfreund** [comment]
@richardpringle here's the sandbox repo https://github.com/michaelfreund/loopback-sandbox-issue-2220. Just open the explorer and try GET /media/{id}/author. Since there is no data, any value for id should work to reproduce the issue.

**Comment by richardpringle** [comment]
@mediastuttg @michaelfreund (which one should I use? haha), you're right!

What happens is a `findById` call is triggered before the actual remote method is called and the error is returned right away instead of passed along to the `__get__author` function.

@bajtos, do you think it's possible to re-wire strong-remoting such that the error propagates down the middleware chain (in which case ACL would kick in) and then the 401 error would override the 404? That way remote hooks would still work even if an error is thrown beforehand. I'm not sure how much re-factoring would be required to accomplish this though.

@mediastuttgart, in the meantime, I would suggest that you work with custom middleware function to check which remote method has being called and handle access rights appropriately from there. How you do that is dependent on what type of ACL you really want (assuming you don't actually want to deny everyone from accessing the method).

@bajtos, if you have a better suggestion for a workaround, please chime in :smile:.

**Comment by richardpringle** [comment]
side note: 
If you do not include an ACL rule on `Media.findById` then I would expect that a user would be able to get the 404 because they could just as easily check `GET /media/{id}` before calling `GET /media/{id}/author`.

However, I added the ACL to deny-all for `findById` and you still run into the same issue. Just thought I would point this out anyway.

**Comment by bajtos** [comment]
I suspect this may have the same cause as https://github.com/strongloop/loopback/issues/2398 and that's the way how strong-remoting handles prototype methods. It always calls `sharedCtor` (which is implemented via find/findById) before any ACL checks are made. See [restPrototypeHandler](https://github.com/strongloop/strong-remoting/blob/b03dfe8ec53d4ca694a2e438d6b5dadff0824305/lib/rest-adapter.js#L446-L455).

I don't know any workaround from the top of my head, I think this needs to be fixed in strong-remoting.

**Comment by bajtos** [comment]
@richardpringle it may be better to create a new issue describing the root problem (prototype methods apply ACL checks only after the model instance was found) and close #2220 and #2398 as duplicates.

**Comment by richardpringle** [comment]
@mediastuttgart, I am closing this issue in favor of strongloop/strong-remoting#311 as the fix needs to be made in that repository.

Thank you for your contribution, and if I were you I would subscribe to the issue to see when it is solved. Also feel free to leave any comments in there, especially if you come up with a workaround for now (I can also try to help you with that if you have any questions).

**Comment by michaelfreund** [comment]
@richardpringle thanks. I will subscribe to strongloop/strong-remoting#311. thanks for your suggestion.
